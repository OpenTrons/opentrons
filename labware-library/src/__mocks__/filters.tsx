'use strict'

jest.mock('../definitions')

const filters = jest.genMockFromModule('../filters')

// commonjs export to mock named exports
module.exports = filters
