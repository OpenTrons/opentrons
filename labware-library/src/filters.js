// @flow
// filter helpers
import queryString from 'query-string'
import pickBy from 'lodash/pickBy'
import uniq from 'lodash/uniq'

import { getAllDefinitions } from './definitions'
import { getPublicPath } from './public-path'

import type { Location } from 'react-router-dom'
import type { FilterParams } from './types'

export const FILTER_OFF = 'all'

export function getAllCategories(): Array<string> {
  const categories = getAllDefinitions().map(d => d.metadata.displayCategory)

  return [FILTER_OFF].concat(uniq(categories))
}

export function getAllManufacturers(): Array<string> {
  const manufacturers = getAllDefinitions().map(d => d.brand.brand)

  return [FILTER_OFF].concat(uniq(manufacturers))
}

export function getFilters(location: Location): FilterParams {
  const queryParams = queryString.parse(location.search)
  const category = queryParams.category || FILTER_OFF
  const manufacturer = queryParams.manufacturer || FILTER_OFF

  return { category, manufacturer }
}

export function buildFiltersUrl(filters: FilterParams): string {
  const params = pickBy(filters, v => v !== FILTER_OFF)

  return `${getPublicPath()}?${queryString.stringify(params)}`
}
