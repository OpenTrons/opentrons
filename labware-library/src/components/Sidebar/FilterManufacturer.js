// @flow
// filter labware by manufacturer
import * as React from 'react'
import { withRouter } from 'react-router-dom'

import { SelectField } from '@opentrons/components'
import { getAllManufacturers, buildFiltersUrl } from '../../filters'
import styles from './styles.css'

import type { ContextRouter } from 'react-router-dom'
import type { FilterParams } from '../../types'

// TODO(mc, 2019-03-13): i18n
const EN_MANUFACTURER = 'manufacturer'

// TODO(mc, 2019-03-18): i18n
const EN_MANUFACTURER_LABELS = {
  all: 'All',
  generic: 'Generic',
}

export type FilterManufacturerProps = {
  ...ContextRouter,
  filters: FilterParams,
}

export function FilterManufacturer(props: FilterManufacturerProps) {
  const { history, filters } = props
  const manufacturers = getAllManufacturers()
  const options = manufacturers.map(value => ({
    value,
    label: EN_MANUFACTURER_LABELS[value] || value,
  }))

  return (
    <label className={styles.filter_manufacturer}>
      <p className={styles.filter_manufacturer_label}>{EN_MANUFACTURER}</p>
      <SelectField
        className={styles.filter_manufacturer_select}
        name="manufacturer"
        value={filters.manufacturer}
        options={options}
        onValueChange={(_, value) => {
          if (value) {
            history.push(buildFiltersUrl({ ...filters, manufacturer: value }))
          }
        }}
      />
    </label>
  )
}

export default withRouter(FilterManufacturer)
