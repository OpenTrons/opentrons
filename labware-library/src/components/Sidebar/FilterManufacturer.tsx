// filter labware by manufacturer
import { getAllManufacturers, buildFiltersUrl } from '../../filters'
import { MANUFACTURER, MANUFACTURER_VALUES } from '../../localization'
import type { FilterParams } from '../../types'
import styles from './styles.css'
import { SelectField } from '@opentrons/components'
import type { SelectOptionOrGroup } from '@opentrons/components'
import * as React from 'react'
import { withRouter } from 'react-router-dom'
import type { RouteComponentProps } from 'react-router-dom'

export interface FilterManufacturerProps extends RouteComponentProps {
  filters: FilterParams
}

export function FilterManufacturerComponent(
  props: FilterManufacturerProps
): JSX.Element {
  const { history, filters } = props
  const manufacturers = getAllManufacturers()
  const options: SelectOptionOrGroup[] = manufacturers.map(value => ({
    value,
    label: MANUFACTURER_VALUES[value] || value,
  }))

  options.push({ options: [{ label: 'Other', value: 'other' }] })

  return (
    <label className={styles.filter_manufacturer}>
      <p className={styles.filter_label}>{MANUFACTURER}</p>
      <SelectField
        className={styles.filter_manufacturer_select}
        name="manufacturer"
        value={filters.manufacturer}
        options={options}
        onValueChange={(_, value) => {
          if (value) {
            history.push(buildFiltersUrl({ ...filters, manufacturer: value }))
          }
        }}
      />
    </label>
  )
}

export const FilterManufacturer = withRouter(FilterManufacturerComponent)
