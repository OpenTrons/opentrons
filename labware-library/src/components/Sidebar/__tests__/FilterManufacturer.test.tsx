// FilterManufacturer component tests
import * as React from 'react'
import { shallow } from 'enzyme'

import { FilterManufacturerComponent } from '../FilterManufacturer'
import * as filters from '../../../filters'

jest.mock('../../../filters')

const getAllManufacturers = filters.getAllManufacturers as jest.MockedFunction<
  typeof filters.getAllManufacturers
>

describe('FilterManufacturer', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('component renders', () => {
    getAllManufacturers.mockReturnValue(['all', 'foo', 'bar', 'baz'])

    const filters = { category: 'all', manufacturer: 'all' }
    const tree = shallow(
      <FilterManufacturerComponent
        location={{} as any}
        history={{} as any}
        match={{} as any}
        filters={filters}
      />
    )

    expect(tree).toMatchSnapshot()
  })
})
