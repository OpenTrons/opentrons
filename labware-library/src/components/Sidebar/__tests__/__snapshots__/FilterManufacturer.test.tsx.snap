// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterManufacturer component renders 1`] = `
<label
  className="filter_manufacturer"
>
  <p
    className="filter_label"
  >
    manufacturer
  </p>
  <SelectField
    className="filter_manufacturer_select"
    name="manufacturer"
    onValueChange={[Function]}
    options={
      Array [
        Object {
          "label": "All",
          "value": "all",
        },
        Object {
          "label": "foo",
          "value": "foo",
        },
        Object {
          "label": "bar",
          "value": "bar",
        },
        Object {
          "label": "baz",
          "value": "baz",
        },
        Object {
          "options": Array [
            Object {
              "label": "Other",
              "value": "other",
            },
          ],
        },
      ]
    }
    value="all"
  />
</label>
`;
